sdks/java/io/pulsar/src/main/java/org/apache/beam/sdk/io/pulsar/ReadFromPulsarDoFn.java

```diff
@@ -0,0 +1,219 @@
+package org.apache.beam.sdk.io.pulsar;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.range.OffsetRange;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.splittabledofn.*;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.apache.pulsar.client.admin.PulsarAdminException;
+import org.apache.pulsar.client.api.*;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Supplier;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Suppliers;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.*;
+
+@DoFn.UnboundedPerElement
+@SuppressWarnings("rawtypes")
+public class ReadFromPulsarDoFn extends DoFn<PulsarSourceDescriptor, PulsarMessage> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReadFromPulsarDoFn.class);
+    private PulsarClient client;
+    private PulsarAdmin admin;
+    private String clientUrl;
+    private String adminUrl;
+
+    @VisibleForTesting Reader<byte[]> readerTst;
```

Comment:
Can we avoid writing code that just enables testing, and do mocking instead?

Reply:
As you can see in the SDF, I made the connection with Pulsar client and then create a subscription with the topic through the reader. So in the unit testing of the reader SDF I use a mock reader, which I need to pass in the SDF so instead of make a connection with pulsar it uses the mock reader, as you can see in `ReadFromPulsarDoFnTest` in `@Setup`. Is there any suggestion on how can I avoid this so i don't have any code for test in main class?

Reply:
Generally, it is better to inject dependencies like this, instead of constructing them within the method. In this case, I would look to mock the PulsarClient, so when you call .newReader on line 111 it returns the fake reader instead.

Reply:
Already made some changes in order to avoid injecting test code into the main class, I take as an example how has been implemented in KafkaIO. Let me know your thoughts!